### PR TITLE
feat: Make `ui` a Python package extra

### LINKED ISSUE #7006: Make `ui` a Python package extra
@z3z1ma pointed out that removing the UI libraries significantly decreases the size of the meltano install. 

It could impact our tests and might take a while to get right, but it probably is worth considering.

### PR DESCRIPTION
**Pull request by rawwar**
Fix #7006

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -68,18 +68,18 @@ click-default-group = "^1.2.1"
...
-flask-sqlalchemy = "^2.4.4"
+flask = { version = "^2.1", optional = true }
+flask-cors = { version = "^3.0.7", optional = true }
+flask-executor = { version = "^0.10", optional = true }
+flask-login = { version = "==0.6.1", optional = true }  # https://github.com/meltano/meltano/issues/6548
+flask-restful = { version = "^0.3.7", optional = true }
+flask-sqlalchemy = { version = "^2.4.4", optional = true }
 flatten-dict = "^0"
 google-cloud-storage = {version = ">=1.31.0", optional = true}
 gunicorn = "^20.1.0"
```
Comment (WillDaSilva) [inline pyproject.toml:79]:
```suggestion
gunicorn = { version = "^20.1.0", optional = true }
```

File: pyproject.toml [hunk 2]
```diff
@@ -100,13 +100,14 @@ structlog = "^23.1.0"
...
 virtualenv = "^20.23.0"
-werkzeug = ">=2.1,<=2.1.3"
+werkzeug = {version= ">=2.1,<=2.1.3", optional = true}
 
 [tool.poetry.extras]
 azure = ["azure-storage-blob", "azure-common", "azure-core"]
 gcs = ["google-cloud-storage"]
 mssql = ["pymssql"]
 s3 = ["boto3"]
+ui = ["flask", "flask-cors", "flask-executor", "flask-login", "flask-restful", "flask-sqlalchemy", "meltano-flask-security", "werkzeug"]
```
Comment (WillDaSilva) [inline pyproject.toml:110]:
```suggestion
ui = ["flask", "flask-cors", "flask-executor", "flask-login", "flask-restful", "flask-sqlalchemy", "gunicorn", "meltano-flask-security", "werkzeug"]
```

**Review by WillDaSilva — COMMENTED** [review]
Thanks for tackling this @rawwar! In addition to the dependencies you've marked as optional, I think we can also include `uvicorn`, `email-validator`, `bcrypt`, `authlib`, and `gunicorn` (see PR suggestions below). Additionally, `backoff` can be made into a dev dependency.

We'll also need to ensure that all Meltano commands function properly except for the `ui` and `user` commands, which should produce an error message stating that the `ui` extra must be installed to use that functionality.

The Meltano Docker image defined within this repository should be updated to install the `ui` extra, to avoid breaking changes prior to Meltano v3.

**Comment by rawwar** [comment]
Thanks @WillDaSilva . I'll keep them in mind.

**Comment by jlloyd-widen** [comment]
@rawwar Any more progress on this? Happy to help if I can.

**Comment by jlloyd-widen** [comment]
In case anyone is interested in taking a look. I have a very rough set of code that seems to pull this off 
https://github.com/jlloyd-widen/meltano/tree/feature/ui-extra 

I made it as a PR to @rawwar 's branch. 

It does not add the `ui` extra to the set of tests however. Nor have I addressed the Docker image. I was unsure of the way to approach either if someone wants to make a suggestion...

**Comment by edgarrmondragon** [comment]
> It does not add the `ui` extra to the set of tests however. Nor have I addressed the Docker image. I was unsure of the way to approach either if someone wants to make a suggestion...

Thanks @jlloyd-widen! For the tests part, you might wanna take a look at how those are handled in the noxfile: https://github.com/meltano/meltano/blob/20f38e75d07d9900b077fa4609e4228ff4eccac3/noxfile.py#L37-L41

As for the Docker image, it's a similar approach: https://github.com/meltano/meltano/blob/20f38e75d07d9900b077fa4609e4228ff4eccac3/docker/meltano/Dockerfile#L25

Does that help?

**Comment by jlloyd-widen** [comment]
@edgarrmondragon Great, thanks! I added that. Now, I think it's a matter of @rawwar merging [my PR](https://github.com/rawwar/meltano/pull/1) into his branch, or I create a new PR directly to this meltano repo ...

**Comment by rawwar** [comment]
@jlloyd-widen  Just merged your pr into the current one

**Comment by edgarrmondragon** [comment]
The UI has been removed altogether in v3. Thanks for the work on this @rawwar!

**Pull request closed without merging**